applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py

```diff
@@ -0,0 +1,548 @@
+        active_constraint_values = []
+        active_constraint_variables = []
+        active_relaxation_coefficient = []
+        active_correction_coefficient = []
+
+        for constraint in self.constraints:
+            if self.__isConstraintActive(constraint):
+                identifier = constraint["identifier"].GetString()
+                constraint_value = self.communicator.getStandardizedValue(identifier)
+                buffer_value = self.constraint_buffer_variables[identifier]["buffer_value"]
+
+                active_constraint_values.append(constraint_value)
+                g_a_variable = self.constraint_gradient_variables[identifier]["mapped_gradient"]
+                g_a_norm = self.optimization_utilities.ComputeMaxNormOfNodalVariable(self.design_surface, g_a_variable)
+                g_a_variable_vector = KM.Vector()
+                self.optimization_utilities.AssembleVector(self.design_surface, g_a_variable_vector, g_a_variable)
+                if abs(g_a_norm) > 1e-10:
+                    g_a_variable_vector *= 1.0/g_a_norm
+
+                active_constraint_variables.append(g_a_variable_vector)
+                active_relaxation_coefficient.append(min(buffer_value,1.0))
+
+                max_buffer = 2.0
+                if buffer_value > 1.0:
+                    if buffer_value < max_buffer:
+                        active_correction_coefficient.append(2*(buffer_value - 1))
+                    else:
+                        active_correction_coefficient.append(2*(max_buffer-1))
+                else:
+                    active_correction_coefficient.append(0.0)
+
+
+        return active_constraint_values, active_constraint_variables, active_relaxation_coefficient, active_correction_coefficient
+
+    # --------------------------------------------------------------------------
+    def __isConstraintActive(self, constraint):
+        identifier = constraint["identifier"].GetString()
+        constraint_value = self.communicator.getStandardizedValue(identifier)
+        if constraint["type"].GetString() == "=":
+            return True
+        elif constraint_value >= self.constraint_buffer_variables[identifier]["lower_buffer_value"]:
+            return True
+        else:
+            return False
+    # --------------------------------------------------------------------------
+    def __updateBufferZone(self):
+        # adapt the buffer zones for zig-zagging, too much or too little correction
+        for constraint in self.constraints:
+
+            identifier = constraint["identifier"].GetString()
+            constraint_value = self.communicator.getStandardizedValue(identifier)
+            constraint_value_m1 = self.constraint_buffer_variables[identifier]["constraint_value-1"]
+            constraint_value_m2 = self.constraint_buffer_variables[identifier]["constraint_value-2"]
+            constraint_value_m3 = self.constraint_buffer_variables[identifier]["constraint_value-3"]
```

Comment:
Is there a possibility to have more descriptive names?

Reply:
g_i_m_1 means g_i-1 ?? basically m_1 means -1 in index.